dm/dm/master/scheduler/scheduler.go

```diff
@@ -693,15 +724,71 @@ func (s *Scheduler) TransferSource(source, worker string) error {
 	if err2 := s.updateStatusToBound(w, ha.NewSourceBound(source, worker)); err2 != nil {
 		s.logger.DPanic("we have checked w.stage is free, so there should not be an error", zap.Error(err2))
 	}
-
-	// 6. try bound the old worker
+	// 6. now this old worker is free, try bound source to it
 	_, err = s.tryBoundForWorker(oldWorker)
 	if err != nil {
 		s.logger.Warn("in transfer source, error when try bound the old worker", zap.Error(err))
 	}
+	s.mu.Unlock()
 	return nil
 }
 
+// batchOperateTaskOnWorker batch operate tasks in one worker and use query-status to make sure all tasks are in expected stage if needWait=true.
+func (s *Scheduler) batchOperateTaskOnWorker(
+	ctx context.Context, worker *Worker, tasks []string, source string, stage pb.Stage, needWait bool) error {
+	for _, taskName := range tasks {
+		if err := s.UpdateExpectSubTaskStage(stage, taskName, source); err != nil {
+			return err
+		}
+	}
+	if !needWait {
+		return nil
+	}
+	// wait all tasks are in expected stage before actually starting scheduling
+WaitLoop:
+	for retry := 0; retry < maxQueryWorkerRetryNum; retry++ {
+		resp, err := worker.queryStatus(ctx)
+		if err != nil {
+			return terror.Annotatef(err, "failed to query worker: %s status", worker.baseInfo.Name)
+		}
+
+		failpoint.Inject("batchOperateTaskOnWorkerMustRetry", func(v failpoint.Value) {
+			if retry < v.(int) {
+				resp.QueryStatus.SubTaskStatus[0].Stage = pb.Stage_InvalidStage
+				log.L().Info("batchOperateTaskOnWorkerMustRetry failpoint triggered", zap.Int("retry", retry))
+			} else {
+				log.L().Info("batchOperateTaskOnWorkerMustRetry passed", zap.Int("retry", retry))
+			}
+		})
+
+		for _, status := range resp.QueryStatus.GetSubTaskStatus() {
+			if status.Stage != stage {
```

Comment:
Is that possible for `status` to be `nil`?

Reply:
it should not happend when our server side rpc not changed, but i add a check in https://github.com/pingcap/tiflow/pull/3889/commits/f43c16e334951d6a0784b15b872f0146f957203a